storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
 
 	mu  sync.Mutex
 	err error
+
+	stream storagepb.Storage_WriteObjectClient
 }
 
-func (w *Writer) open() error {
-	attrs := w.ObjectAttrs
-	// Check the developer didn't change the object Name (this is unfortunate, but
-	// we don't want to store an object under the wrong name).
-	if attrs.Name != w.o.object {
-		return fmt.Errorf("storage: Writer.Name %q does not match object name %q", attrs.Name, w.o.object)
+// openGRPC initializes a pipe for the user to write data to, and a routine to
+// read from that pipe and upload the data to GCS via gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) openGRPC() error {
+	if err := w.validateWriteAttrs(); err != nil {
+		return err
 	}
-	if !utf8.ValidString(attrs.Name) {
-		return fmt.Errorf("storage: object name %q is not valid UTF-8", attrs.Name)
+
+	pr, pw := io.Pipe()
+	w.pw = pw
+	w.opened = true
+
+	go w.monitorCancel()
+
+	// TODO: Handle this case later.
+	bufSize := w.ChunkSize
+	if w.ChunkSize == 0 {
+		bufSize = maxPerMessageWriteSize
 	}
-	if attrs.KMSKeyName != "" && w.o.encryptionKey != nil {
-		return errors.New("storage: cannot use KMSKeyName with a customer-supplied encryption key")
+
+	var offset int64
+	var upid string
+
+	go func() {
+		defer close(w.donec)
+
+		// Loop until there is an error or the Object has been finalized.
+		for {
+			// Initiliaze client buffer with ChunkSize.
+			buf := make([]byte, bufSize)
+			recvd, done, err := read(pr, buf)
+			if err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
+				return
+			}
+
+			// TODO: Figure out how to set up encryption via CommonObjectRequestParams.
+			// TODO: Apply Object write conditions to the request.
+			// TODO: Send Object checksum.
+
+			// The chunk buffer is full, but there is no end in sight. Start a
+			// resumable upload if it has not already been started.
+			if !done && upid == "" {
+				upid, err = w.startResumableUpload()
+				if err != nil {
+					w.error(err)
+					pr.CloseWithError(err)
+					return
+				}
+			}
+
+			o, off, finalized, err := w.upload(buf, recvd, offset, done, upid)
+			if err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
```

Comment:
Well, the desired affect is to close the read side of the pipe when we encounter an error so that the writer knows the upload was aborted. Their Write calls will fail because the pipe was closed. So yes. But I agree in that we need error mode tests to ensure things work as expected.